### PR TITLE
Enable Cpp include scanning through option

### PR DESCRIPTION
**Pull request by werkt**
Provide a --should_scan_includes option to enable include scanning,
which allows, as previously implemented, individual compilation action
inputs to be extended with the contents of make-style (.d) include
dependencies after an initial compile, limiting their recompilation to
explicit dependencies' invalidations, versus those gathered from
target-level deps.

Option description includes admonition of incompatibility with
sandboxing.

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

File: src/main/java/com/google/devtools/build/lib/bazel/rules/cpp/BazelCppSemantics.java [hunk 1]
```diff
@@ -42,12 +42,16 @@ public PathFragment getEffectiveSourcePath(Artifact source) {
...
-    // including header files, as opposed to just the "compile" filegroup.
-    actionBuilder.addTransitiveMandatoryInputs(CppHelper.getToolchain(ruleContext).getCrosstool());
-    actionBuilder.setShouldScanIncludes(false);
+    boolean shouldScanIncludes = cppConfiguration.shouldScanIncludes();
+    if (!shouldScanIncludes) {
+      // When not performing include scanning, we need the entire crosstool filegroup,
+      // including header files, as opposed to just the "compile" filegroup.
+      actionBuilder.addTransitiveMandatoryInputs(CppHelper.getToolchain(ruleContext).getCrosstool());
+    }
+    actionBuilder.setShouldScanIncludes(shouldScanIncludes);
```
Comment (abergmeier-dsfishlabs) [inline src/main/java/com/google/devtools/build/lib/bazel/rules/cpp/BazelCppSemantics.java]: @damienmg Out of curiousity, is scanning of includes using the same `genrule` infrastructure we have available in Skylark or is it cheating?

Comment (damienmg) [inline src/main/java/com/google/devtools/build/lib/bazel/rules/cpp/BazelCppSemantics.java]: I am not too familiar with include scanning but IIRC include scanning is to skip strict header checking to know the actual header used. Not to improve incrementality. This is kind of ugly to do in general.

Comment (lberki) [inline src/main/java/com/google/devtools/build/lib/bazel/rules/cpp/BazelCppSemantics.java]: `genrule` infrastructure? What's that?

include scanning is custom C++ specific logic in Blaze. It serves two purposes: add includes that are not specified by the rule but are still needed (bad) and remove includes from the inputs of the action that are not actually needed (good, helps with action caching especially when remote executors are in play)

**Comment by lberki** [comment]
Ugh, I'd much, much rather not do this because it pokes a hole into the "all headers that are used need to be specified" rule we have. What do you need this for?

I think we'll eventually need to re-enable include scanning so that remote action execution works better, but that should be done in such a way that headers can only be removed and not added.

**Comment by abergmeier-dsfishlabs** [comment]
> Ugh, I'd much, much rather not do this because it pokes a hole into the "all headers that are used need to be specified" rule we have.

That's what I meant with cheating :)

**Comment by werkt** [comment]
> Ugh, I'd much, much rather not do this because it pokes a hole into the "all headers that are used need to be specified" rule we have.

Except that header inclusion checking (which cannot be disabled) prevents any header not specified in a dependency's manifest from successfully building.

> What do you need this for?

Build performance.  Headers are transitive through dependencies, which means that, without include scanning, if a target declares headers, and a header it owns is modified, every single translation-unit-member of srcs for every transitive dependent is recompiled, whether it includes the header or not. I refuse to make single-header targets out of my current build - the collections of headers represent meaningful interfaces, and it would make the build definitions unbelievably intractable and unmaintainable.

**Comment by ulfjack** [comment]
We have some code to subset the headers in the action to the set of headers that were actually read by the compiler, and I think that's reasonable for us to allow with an option (though note that we've found correctness bugs in MSVC), or even by default. However, it's not what this code change does, as far as I can tell. It's also not what we internally call include scanning.

Note: what we internally call include scanning isn't necessarily what external users understand.

**Comment by ulfjack** [comment]
Let me try to clarify: we have code that updates the inputs of an C++ compile action after execution to remove inputs that were not used during that execution, such that changes to unused headers do not trigger re-execution of that action. This is compatible with both sandboxing and remote execution. We should have a flag to allow enabling that code, if we don't already (it's possible that we have a flag already). That seems like it addresses @werkt's concerns. Do you agree?

**Comment by werkt** [comment]
Yes, it would.  I will leave it to you to provide this code and option, and this request can be closed when that has landed.  Timeframe?  0.5?

**Comment by lberki** [comment]
AFAICT we already do .d file pruning:

https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/rules/cpp/CppCompileAction.java#L1278
https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/bazel/rules/cpp/BazelCppSemantics.java#L68

Do you have evidence that Bazel doesn't do that for some reason? If so, that's a pretty gruesome bug.

**Comment by werkt** [comment]
@lberki You have the capacity, but there is no use of it:

```
    // Note that this effectively disables post-execute "include scanning" in Bazel, because
    // hdrs_check is forced to "strict" and "include scanning" is forced to off.
    boolean usesStrictHdrsChecks = context.getDeclaredIncludeDirs().isEmpty()
        && context.getDeclaredIncludeWarnDirs().isEmpty();
    if (shouldScanIncludes() || !usesStrictHdrsChecks) {
      updateActionInputs(discoveredInputs);
    }
```

`discoveredInputs` is the result of the function you've mentioned.  You cannot enable the effects of this pruning without code modification, because, as the comment indicates, include scanning is not controllable (and I haven't evaluated the other side of the ||, but i trust the comment in saying that it is disabled under `!usesStrictHdrsChecks` as well).

**Comment by lberki** [comment]
I see, I overlooked that one. In theory, it shouldn't be hard to enable this for Bazel then and if it works as expected, it wouldn't even be noticeable (apart from better incrementality)

**Comment by ulfjack** [comment]
Can you redo this change to enable input subsetting, and make sure that it works with sandboxing?

**Comment by werkt** [comment]
I'll see what I can make of it.  I will have to make an assessment of whether to supply a compile with all transitive inputs based on whether it has any inputs already, which I hope will be sufficient.

**Comment by werkt** [comment]
Most recent update enables subsetting and works with sandboxing

File: src/main/java/com/google/devtools/build/lib/bazel/rules/cpp/BazelCppSemantics.java (hunk 1 above)
Comment (ulfjack) [inline src/main/java/com/google/devtools/build/lib/bazel/rules/cpp/BazelCppSemantics.java]: Skylark can't do include scanning in any form or shape, so this is something Skylark rules cannot do right now. If you consider that 'cheating', then yes, it's cheating.

File: src/main/java/com/google/devtools/build/lib/bazel/rules/cpp/BazelCppSemantics.java [hunk 2]
```diff
@@ -42,12 +42,16 @@ public PathFragment getEffectiveSourcePath(Artifact source) {
...
     actionBuilder.setActionContext(CppCompileActionContext.class);
-    // Because Bazel does not support include scanning, we need the entire crosstool filegroup,
-    // including header files, as opposed to just the "compile" filegroup.
-    actionBuilder.addTransitiveMandatoryInputs(CppHelper.getToolchain(ruleContext).getCrosstool());
-    actionBuilder.setShouldScanIncludes(false);
+    boolean shouldScanIncludes = cppConfiguration.shouldScanIncludes();
+    if (!shouldScanIncludes) {
+      // When not performing include scanning, we need the entire crosstool filegroup,
+      // including header files, as opposed to just the "compile" filegroup.
+      actionBuilder.addTransitiveMandatoryInputs(CppHelper.getToolchain(ruleContext).getCrosstool());
```
Comment (ulfjack) [inline src/main/java/com/google/devtools/build/lib/bazel/rules/cpp/BazelCppSemantics.java]: I think this is a no-op for locally installed toolchains on Linux. We need to check if this works with checked-in toolchains or not. I suspect it doesn't, and that we should always add the crosstool (i.e., remove the if above). The subsetting should still work, right? Even if not, subsetting the crosstool headers is probably not critical, unless someone modifies them all the time - is anyone doing that?

**Comment by ulfjack** [comment]
Ideally, we could improve out the terminology before this goes in, and we also need to add some tests. Should we call this --should_scan_includes? For Google internal users, that is going to be confusing, because we usually use the term include scanning to refer to finding header files before running an action rather than subsetting header files afterwards.

File: src/main/java/com/google/devtools/build/lib/bazel/rules/cpp/BazelCppSemantics.java (hunk 2 above)
Comment (werkt) [inline src/main/java/com/google/devtools/build/lib/bazel/rules/cpp/BazelCppSemantics.java]: I will remove the if above.  While we have a number of compiler variation selections, we've done all of the legwork for cross/multi-target products in skylark proper, owing to the poor documentation and constrained observable behavior in CROSSTOOL </shameless-req-for-additional-docs>.

**Comment by werkt** [comment]
I will write some tests, hoping that there's some boilerplate for action execution determination in the existing ones...
The --should_scan_includes name was only based on the member var and the proximity to the desired behavior.  How does --strict-cpp-(include|header)-[action-]inputs strike you?  I will change the option and member name accordingly as well based on the change.

**Comment by ulfjack** [comment]
We do have a bunch of tests, but not all of them are open source yet, unfortunately. If you can't find an existing test that almost covers this already, then someone here will have to write one.

**Comment by ulfjack** [comment]
For the flag name, I'd try to stay closer to what it actually does. Maybe --subset_cpp_headers or --detect_used_cpp_headers.

**Comment by lberki** [comment]
I'll close this pull request, since this issue has been resolved in 840acb1 .

**Pull request closed without merging**